src/amo/components/CollectionManager/index.js

```diff
@@ -54,29 +54,53 @@ type InternalProps = {|
 |};
 
 type State = {|
-  customSlug?: boolean,
-  description?: string | null,
-  name?: string | null,
-  slug?: string | null,
+  collectionId: number | null,
+  customSlug: boolean,
+  description: string,
+  name: string,
+  slug: string,
 |};
 
 export class CollectionManagerBase extends React.Component<
   InternalProps,
   State,
 > {
+  static getDerivedStateFromProps(props: InternalProps, state: State) {
+    if (props.collection && props.collection.id !== state.collectionId) {
+      // Only reset the form when receiving a collection that the user is not
+      // already editing. This prevents clearing the form in a few scenarios
+      // such as pressing the submit button.
+
+      // Decode HTML entities so the user sees real symbols in the form.
+      return {
+        collectionId: props.collection.id,
+        customSlug: false,
+        description: props.collection
+          ? decodeHtmlEntities(props.collection.description)
+          : '',
+        name: props.collection ? decodeHtmlEntities(props.collection.name) : '',
+        slug: props.collection ? props.collection.slug : '',
+      };
+    }
+
+    return null;
+  }
+
   constructor(props: InternalProps) {
     super(props);
-    this.state = this.propsToState(props);
-  }
 
-  componentWillReceiveProps(props: InternalProps) {
-    const existingId = this.props.collection && this.props.collection.id;
-    if (props.collection && props.collection.id !== existingId) {
-      // Only reset the form when receiving a collection that the
-      // user is not already editing. This prevents clearing the form
-      // in a few scenarios such as pressing the submit button.
-      this.setState(this.propsToState(props));
-    }
+    this.state = {
```

Comment:
I see a lot of duplication between this and `getDerivedStateFromProps`. Maybe they can share an intermediary function?

Reply:
Snap!